### PR TITLE
commons #29 Caching Converter: support custom keys

### LINKED ISSUE #29: Caching Converter: support custom keys
The current implementation uses an input value as a caching key, but sometimes it doesn't suffice. For instance when `From` type doesn't have `equal/hashcode` implemented correctly, the number of the input values is too bug, or for any other reason it cannot be used as a key for caching. So a there has to be a way to specify a key function to the `CachingConverter`

### PR DESCRIPTION
**Pull request by wajda**
fixes #29

### CODE REVIEW

**Review by yruslan — APPROVED** [review]

File: src/main/scala/za/co/absa/commons/lang/converters.scala [hunk 1]
```diff
@@ -34,10 +34,16 @@ trait Converter {
...
   def values: Seq[To] = cache.values.toSeq
 
+  protected type Key = Any
+  protected def key(x: From): Key = x
+
   abstract override def convert(arg: From): To =
-    cache.getOrElseUpdate(arg, super.convert(arg))
+    cache.getOrElseUpdate(
+      key(arg),
+      super.convert(arg)
```
Comment (benedeki) [inline src/main/scala/za/co/absa/commons/lang/converters.scala:47]: I just wonder if this does, what I believe you want to do.
I think what you want to do with this trait, is avoid multiple conversion if the conversion happened before, right?
But I think the conversion will still happen, as it's evaluated to be passed as parameter to `getOrElseUpdate`,, isn't it? 🤔

File: src/main/scala/za/co/absa/commons/lang/converters.scala [hunk 2]
```diff
@@ -34,10 +34,16 @@ trait Converter {
   * Memoized converter
   */
 trait CachingConverter extends Converter {
-  private val cache = mutable.LinkedHashMap.empty[From, To]
+  private val cache = mutable.LinkedHashMap.empty[Key, To]
 
   def values: Seq[To] = cache.values.toSeq
 
+  protected type Key = Any
+  protected def key(x: From): Key = x
```
[the remark this answers is not in the document]
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/converters.scala:42]: Then maybe `getKey`? `toXXX` sounds like a conversion while it rather is an extraction or a supplement.

File: src/main/scala/za/co/absa/commons/lang/converters.scala (hunk 1 above)
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/converters.scala:47]: > I think what you want to do with this trait, is avoid multiple conversion if the conversion happened before, right?

Yes.

> But I think the conversion will still happen, as it's evaluated to be passed as parameter to getOrElseUpdate,, isn't it?

No. The second parameter of the `getOrElseUpdate()` method is called by-name, meaning lazily.

Comment (benedeki) [reply src/main/scala/za/co/absa/commons/lang/converters.scala:47]: Ahh, yeah. 👍

**Review by cerveada — APPROVED** [review]

File: src/main/scala/za/co/absa/commons/lang/converters.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/converters.scala]: As agreed offline changed to `keyOf`

**Review by cerveada — APPROVED** [review]

**Review by benedeki — APPROVED** [review]

**Pull request merged**